### DISCUSSION TITLE
How to query via point + radius/distance?

### DISCUSSION THREAD

**Discussion by nguillaumin** [ngageoint/geopackage-java#61] [Q&A] [answered]
Hi,

I was able to use the library to query a geopackage based on a bounding box. I was wondering if there was a way to query features based on point + radius/distance? I saw the `GeometryEnvelope` class but I'm not sure it could be used for this?

Any pointers would be appreciated, thanks!

Nicolas

**Comment by bosborn** [comment, marked as answer]
GeometryEnvelope is just the simple features container, bounds without any projections.

You could expand the bounding box and query for that box.  If your projection/ feature table is already in meters (mercator, web mercator, etc), just expand by the distance in each direction.

```
BoundingBox queryBox = new BoundingBox(
		bbox.getMinLongitude() - meters,
		bbox.getMinLatitude() - meters,
		bbox.getMaxLongitude() + meters,
		bbox.getMaxLatitude() + meters);
```

If your bounding box / feature table is in degrees / WGS84, you could project it to a meter projection, expand it, and query with that projection provided.  [This method](https://github.com/ngageoint/geopackage-core-java/blob/6ac07d757ea2928e26cdfc05975d12126d75c9ff/src/main/java/mil/nga/geopackage/tiles/TileBoundingBoxUtils.java#L915) converts a WG84 bounding box to a Web Mercator bounding box.

Then call a [query method](https://github.com/ngageoint/geopackage-java/blob/5fc0cdd5d65fce822ba90bd5f0e71a8b873c5fa5/src/main/java/mil/nga/geopackage/features/index/FeatureIndexManager.java#L2542) with both bounds and projection.

```
FeatureIndexManager indexManager = new FeatureIndexManager(geoPackage, featureTable);
FeatureIndexResults results = indexManager
		.query(queryBox, ProjectionFactory
			.getProjection(ProjectionConstants.EPSG_WEB_MERCATOR));
try {
  for (FeatureRow featureRow : results) {
	  // ...
  }
} finally {
  results.close();
}
```

We [do similar things in our apps](https://github.com/ngageoint/geopackage-android-map/blob/bae42ac47829f08fe85265ac2a56838e094f44eb/geopackage-map/src/main/java/mil/nga/geopackage/map/MapUtils.java#L258), but calculate based on screen pixel bounds.  After the initial query, we sometimes [do fine filtering](https://github.com/ngageoint/geopackage-android-map/blob/bae42ac47829f08fe85265ac2a56838e094f44eb/geopackage-map/src/main/java/mil/nga/geopackage/map/features/FeatureInfoBuilder.java#L763) on the results.

**Reply by nguillaumin** [reply]
Thanks for the reply!

This is still ultimately a bounding box which is a "square" area, correct? In my case I need a circle, i.e. I want to query features from within X meters radius of a specific point. Is there a way to achieve this?

**Reply by bosborn** [reply]
Would still be a square.  The GeoPackage RTree index only tracks min/max x/y values.

That's where the fine filtering on the matching results could be done after the fact.  Query for the square and fine filter each matched feature to see if any part of the geometry fits in the "in-circle".

There are some geometry utilities [here](https://github.com/ngageoint/simple-features-java/blob/master/src/main/java/mil/nga/sf/util/GeometryUtils.java) and tile utilities [here](https://github.com/ngageoint/geopackage-core-java/blob/master/src/main/java/mil/nga/geopackage/tiles/TileBoundingBoxUtils.java).

**Reply by nguillaumin** [reply]
Ok, thanks for your help! I will try this.
